### ISSUE TITLE
Parse settings from pyproject.toml

### ISSUE THREAD

**Issue by miigotu** [python-babel/babel#777]
This project is the last of my dependencies that can only read from setup.cfg. the domains that I think should be used are something like:
```toml
;pyproject.toml
[ tool.babel.extract_messages ]
message_extractors="""{
    'gui': [
        ('**/js/*.min.js', 'ignore', None),
        ('**/js/*.js', 'javascript', {'input_encoding': 'utf-8'})
    ]}"""
;options
[ tool.babel.compile_catalog ]
;options
[ tool.babel.init_catalog ]
;options
[ tool.babel.update_catalog ]
;options
```

**Comment by miigotu** [comment]
@akx @benselme @blagasz @kvesteri @mitsuhiko @sils 
Sorry for the ping but I'd really like to see this, and I'm willing to add a PR to make this happen. Before I waste my time, is there a policy or stopper against accepting a change pr that accomplishes this? I just don'y want to waste my time if so.

**Comment by miigotu** [comment]
I'd also like to submit a patch for the f-string problem. There is not a nice way to do it, but I think we could add a hack for it (which includes a small speed expense) that depends on a config/pyproject setting or environment variable or something to enable. Possibly with a new/different import when you gettext_install. That way it would not be a breaking change, and would avoid slowing down every implementation unless they enabled the feature.

**Comment by akx** [comment]
@miigotu There's no policy or stopper preventing a PR for pyproject.toml.

However: 
* care should be taken to keep Babel working even when no TOML capabilities are present,
* the format of the TOML configuration should be well-thought-out in advance. The draft in the original message here seems a bit icky with a Python literal being embedded in the TOML.

As for "the f-string problem"... I think that's quite unrelated here?

**Comment by miigotu** [comment]
Yeah I messed with the spec for it a bit last night, and pulled a global options out. Toml allows multiple definitions of the same key, so 
[tool.babel.config.extractors]
type = python
match = "**/source/*.py"

tool.babel.config.extractors]
type = js
match = "**/js/*.js
```
Works, but I need to think about it more.

**Comment by miigotu** [comment]
Ideally nobody should have to add a secondary config file to their project when pyproject.toml is the preferred and python team recommended method for new projects, and handles all possible use cases in the spec for any tools that is willing to implement their config from it. It was sort of the purpose of the pep, to make it so we didn't need 10 different files for configuration cluttering up our source code, and searching all over the place for where settings came from in larger projects. Setuptools is being phased out in most projects. I understand keeping backwards compatibility, but forwards compatibility keeps tools relevant for people who are perfectionists and prefer some level of interoperability and conformity to standards.

**Comment by jpmckinney** [comment]
> Setuptools is being phased out in most projects

? What are people using to build packages if not setuptools? (enscons?)

**Comment by AbdealiJK** [comment]
I am moving a package to use [poetry](https://github.com/python-poetry/poetry) - I pinged about the config file because of this

I also use basic [build](https://pypi.org/project/build) in some projects

**Comment by miigotu** [comment]
> > Setuptools is being phased out in most projects
> 
> ? What are people using to build packages if not setuptools? (enscons?)

Poetry or flit are the leading tools to package new projects.

Peps 518, 518, 621, 631, 633, 660 (and more) provide for pyproject.toml to include project metadata, build system information, dependencies (including extras and optional named groups), configuration for third party tools and requirements in the expanded toml, elimination for the need of requirements.txt, requirements-dev.txt, requirements-win.txt, setup.py, setup.cfg, and numerous other files in a project. Ideally, a project should only need a readme, pyproject.toml, code, and possibly contributing.md and a LICENSE file. Everything else can be handled in the pyproject.toml provided the tools you use read configuration from it.

I use about a dozen tools in my project, python-babel is the only one that does not check the pyproject except for .editorconfig.

https://github.com/python-poetry/poetry
https://flit.pypa.io/en/stable/

These eliminate the need for multiple tools to build, package, and publish separately. Adding/removing depends, installing (editable or not), multiple virtualenv functionality for testing or building, building/packaging, and pushing to pypi are all handled with one tool. 

Particularly with poetry, you dont need to use virtualenv, pipenv, tox, twine, pip, setuptools, pyenv, or anything. You can even define custom commands to run, inside the pyproject, using poethepoet. It's a complete tool, that is extensible with plugins.

**Comment by jpmckinney** [comment]
FWIW flake8 also doesn't support pyproject.toml https://github.com/PyCQA/flake8/issues/234

Anyway, this project is open to a PR so why not just contribute one? https://github.com/python-babel/babel/issues/777#issuecomment-1258074187

**Comment by akx** [comment]
> Poetry or flit are the leading tools to package new projects.

I disagree - I've been starting to use PyPA's own [`hatch`](https://hatch.pypa.io/latest/) for my libraries as that's what [the official tutorial](https://packaging.python.org/en/latest/tutorials/packaging-projects/) primarily suggests, and `build` to do the building.

**Comment by miigotu** [comment]
> FWIW flake8 also doesn't support pyproject.toml https://github.com/PyCQA/flake8/issues/234
> 
> Anyway, this project is open to a PR so why not just contribute one? https://github.com/python-babel/babel/issues/777#issuecomment-1258074187
> 
> Edit: Off-topic, but I use the pyenv-virtualenv Homebrew package, which automatically installs pip and setuptools, and pip-tools for my applications (not my packages). Then I use the https://github.com/pypa/gh-action-pypi-publish GitHub Action, which uses twine to publish the package. Don't need pipenv (alternative to pip-tools) or tox (just create a matrix in GitHub Actions). I guess one CLI with many packages as plugins might be preferred to multiple CLIs from many packages... Anyway, +1 to pyproject.toml support – I prefer to have fewer config files, too.

Even though most projects are using something like black/blackd instead of flake8, flake8 does in fact support pyproject.toml
https://github.com/SickChill/sickchill/blob/master/pyproject.toml#L125

**Comment by miigotu** [comment]
> > Poetry or flit are the leading tools to package new projects.
> 
> I disagree - I've been starting to use PyPA's own [`hatch`](https://hatch.pypa.io/latest/) for my libraries as that's what [the official tutorial](https://packaging.python.org/en/latest/tutorials/packaging-projects/) primarily suggests, and `build` to do the building.

What are you disagreeing with? Flit and poetry are the two leading projects for pyproject based projects. I never said they are the ONLY options. Even hatch uses pyproject only.

**Comment by miigotu** [comment]
> FWIW flake8 also doesn't support pyproject.toml https://github.com/PyCQA/flake8/issues/234
> 
> Anyway, this project is open to a PR so why not just contribute one? https://github.com/python-babel/babel/issues/777#issuecomment-1258074187
> 
> Edit: Off-topic, but I use the pyenv-virtualenv Homebrew package, which automatically installs pip and setuptools, and pip-tools for my applications (not my packages). Then I use the https://github.com/pypa/gh-action-pypi-publish GitHub Action, which uses twine to publish the package. Don't need pipenv (alternative to pip-tools) or tox (just create a matrix in GitHub Actions). I guess one CLI with many packages as plugins might be preferred to multiple CLIs from many packages... Anyway, +1 to pyproject.toml support – I prefer to have fewer config files, too.

And as far as your edit, I was listing projects that you absolutely have no need for with flit or poetry, because the functionality of all of them is included. I also publish with GitHub actions (again, there's a thousand ways to do this)

The point is that setup.py and setup.cfg are antiquated ways of managing a project, especially if you only have one tool requiring a setup.cfg or .babelrc, when there is a file specifically created and extended for this purpose of configuration of 3rd party tools.

Open source software is always going to have many ways to do things. Our job is to make it easier to do them, and simplify the process, to foster future growth of developers and make their life easier while using our software (and the language we are using).

Just because something CAN be done a certain way, doesn't mean it is the best way. "Hacking" is literally the process of finding things that can be improved and then improving them. When I get around to this, I'll try to make the PR. Unless someone more familiar with this projects source and coding conventions can get to it before then.

**Comment by miigotu** [comment]
> FWIW flake8 also doesn't support pyproject.toml https://github.com/PyCQA/flake8/issues/234
>

Apologies, I use flake9 (fork of flake8) that keeps rebased on flake8 and accepts configuration from pyproject.toml, because of their similar reluctance to accept widely accepted PEPs.

Silly to me to not accept PEPs when the entire project is a result of a PEP specifying code formatting 😂 Afaik, the initial refusal was due to a pip bug when pyproject was initially implemented, that was fixed 2 years ago. Flake8 itself might even work with the pyproject config now, but I have not personally tried it since flake9 has continued to stay up to date and all of the flake8 plugins usually have flake9 counterparts, such as flake9-isort.

I don't think that reluctance to adhere to a new configuration format should be a reason to cause a forking event, but some people have ideas in their head sometimes that are hard to change, myself included.

I just need to find the time away from work and all of my projects to attempt the PR here. It's my birthday today (40 ugh) so commenting is all I'm going to do for today lol.

**Comment by akx** [comment]
@miigotu Happy birthday! 🍰 

Also, yeah, "I disagree" may have been a bit inaccurate - what I meant was that I don't think `poetry` and `flit` are necessarily the two most popular tools, [and `setuptools` itself also does building based on TOML just fine](https://setuptools.pypa.io/en/latest/build_meta.html#how-to-use-it)...

**Comment by miigotu** [comment]
> @miigotu Happy birthday! 🍰 
> 
> Also, yeah, "I disagree" may have been a bit inaccurate - what I meant was that I don't think `poetry` and `flit` are necessarily the two most popular tools, [and `setuptools` itself also does building based on TOML just fine](https://setuptools.pypa.io/en/latest/build_meta.html#how-to-use-it)...

Thanks. So yeah, setuptools does work with pyproject. Setuptools is not really recommended even by pyca or any of the other major groups anymore, but the true successor has not emerged just yet. The few build systems that have came about all have their pros and cons, but that's kind of what the purpose was, for options in how you want your project to be built and distributed.

Poetry might be the best option for rust and c extensions once the documentation is improved for that part, but it's definitely a fast changing environment for build and deployment ecosystems at the moment. Heck I might even release a project to work around f-strings in a nice way, and have it disabled unless someone absolutely needs translations because of the performance loss. 

Point being there is no reason to continue being one of the last major projects that requires a setup.{cfg,py} or a separate configuration file. It's much nicer to have all of that configuration in one place.

**Comment by ikus060** [comment]
Sorry to bring that up again. It's a old issue but still un-resolved. 

What is the roadmap to support the new `pyproject.toml` ? I'm using `python setup.py extract_messages` and other babel integration with setuptools in my project. I'm looking for a way to migrate it to `pyproject.toml`. If even possible.

**Comment by rossnomann** [comment]
@ikus060  You can use `babel.cfg` as a mapping file:

```
[python: **.py]
encoding = utf-8

[jinja2: **.jinja2.html]
encoding = utf-8
```

And run pybabel directly. I use a task runner plugin for poetry:

```
[tool.poe.tasks.xgettext]
cmd = "pybabel extract -F babel.cfg --input-dirs=path/to/project -o path/to/locale/messages.pot"
```

**Comment by ThomasWaldmann** [comment]
Maybe there could be a quick hack until someone wants to do a bigger PR:

https://tox.wiki/en/latest/config.html#pyproject-toml

Something like tox has done there would already enable people to get rid of babel.cfg in their toplevel directory.

**Comment by jpmckinney** [comment]
This works fine for me:

```toml
[project.entry-points."babel.extractors"]
tag = "my.module:extract_mystuff"
```

**Comment by akx** [comment]
@jpmckinney Hi there! Since Babel 2.16 (brand new), you should be able to put the mapping configuration in `pyproject.toml` too; see the syntax over at #1108. Would be grand if you gave it a shot in a Real Life project :)

**Comment by jpmckinney** [comment]
@akx What does the configuration look like? There are many approaches discussed in the PR, and I can see the (supposedly working) TOML example in the PR's tests, but if I do: ``[[babel.mappings]`` I get `babel.messages.frontend.ConfigurationError: pyproject.toml: No 'tool.babel' section found in file`, and if I do `[[tool.babel.mappings]]`, then I get no errors but I also get no messages extracted.

**Comment by blaise-io** [comment]
@akx did https://github.com/python-babel/babel/pull/1108 solve this?

**Comment by ThomasWaldmann** [comment]
@blaise-io I saw your comment and gave it a try. The main problem was the lack of documentation and even after reading the issue and pr it wasn't clear how the toml shall look like (what was implemented in the end).

I did quite some trial and error and ended up with this hopefully correct pyproject.toml entries:

https://github.com/moinwiki/moin/pull/1996/files

Maybe just put a similar config into the docs?
